### PR TITLE
FateIT Repo Error Test

### LINKED ISSUE #4266: Add test to FateIT that tests errors
FateIT needs to test that errors are handled as expected and undo is called.  This could be done with a test like the following

 1.  Fate is seeded with Repo1
 2. Repo1 is called and returns Repo2
 3. Repo2 is called and return Repo3
 4. Repo3 is called and throws an exception (need to test throwing an exception in isReady() and call(), may need to run test twice with that minor diff)

After the above happens fate should call undo() on Repo3, then Repo2, then Repo1.  The test needs to verify that undo is called in the expected order.

It may be worthwhile to add this test to 2.1 and merge it forward.

### PR DESCRIPTION
**Pull request by kevinrr888**
Closes #4266
This adds a test `testRepoFails()` to `FateIT` in 2.1 which ensures `undo()` is called and in the expected order when an error occurs in a Repo's `call()` or `isReady()` methods.

### CODE REVIEW

File: test/src/main/java/org/apache/accumulo/test/fate/zookeeper/FateIT.java [hunk 1]
```diff
@@ -369,6 +429,69 @@ public void testCancelWhileInCall() throws Exception {
...
+      /*
+       * Test exception in isReady()
+       */
+      TestOperationFails.undoOrder = new ArrayList<>();
+      undoLatch = new CountDownLatch(TestOperationFails.TOTAL_NUM_OPS);
+      txid = fate.startTransaction();
+      assertEquals(NEW, getTxStatus(zk, txid));
+      fate.seedTransaction("TestOperationFails", txid,
+          new TestOperationFails(1, ExceptionLocation.IS_READY), true, "Test Op Fails");
+      assertEquals(SUBMITTED, getTxStatus(zk, txid));
```
Comment (keith-turner) [inline test/src/main/java/org/apache/accumulo/test/fate/zookeeper/FateIT.java:485]: There is a race condition with this check, if a sleep is added and the test run will probably see it.  The status could change from SUBMITTED at any time as background fate threads start working on the seeded transaction.

```suggestion
     Thread.sleep(1000);
      assertEquals(SUBMITTED, getTxStatus(zk, txid));
```

File: test/src/main/java/org/apache/accumulo/test/fate/zookeeper/FateIT.java [hunk 2]
```diff
@@ -369,6 +429,69 @@ public void testCancelWhileInCall() throws Exception {
...
+      // Wait for all the undo() calls to complete
+      undoLatch.await();
+      assertEquals(expectedUndoOrder, TestOperationFails.undoOrder);
+      assertEquals(FAILED_IN_PROGRESS, getTxStatus(zk, txid));
```
Comment (keith-turner) [inline test/src/main/java/org/apache/accumulo/test/fate/zookeeper/FateIT.java:489]: There is a race condition with this check, could do the following which should not have a race condition and utilizes and verifies more of the Fate API.  To do the following would need to disable auto cleanup, will make another comment about that.

```suggestion
      assertEquals(FAILED, fate.waitForCompletion(txid));
      assertTrue(fate.getException(txid).getMessage().contains("isReady() failed"));
```

File: test/src/main/java/org/apache/accumulo/test/fate/zookeeper/FateIT.java [hunk 3]
```diff
@@ -369,6 +429,69 @@ public void testCancelWhileInCall() throws Exception {
...
+      /*
+       * Test exception in isReady()
+       */
+      TestOperationFails.undoOrder = new ArrayList<>();
+      undoLatch = new CountDownLatch(TestOperationFails.TOTAL_NUM_OPS);
+      txid = fate.startTransaction();
+      assertEquals(NEW, getTxStatus(zk, txid));
+      fate.seedTransaction("TestOperationFails", txid,
+          new TestOperationFails(1, ExceptionLocation.IS_READY), true, "Test Op Fails");
```
Comment (keith-turner) [inline test/src/main/java/org/apache/accumulo/test/fate/zookeeper/FateIT.java:484]: Disabling auto cleanup makes waitForCompletion suggestion in another comment work.

```suggestion
          new TestOperationFails(1, ExceptionLocation.IS_READY), false, "Test Op Fails");
```

File: test/src/main/java/org/apache/accumulo/test/fate/zookeeper/FateIT.java [hunk 4]
```diff
@@ -369,6 +429,69 @@ public void testCancelWhileInCall() throws Exception {
...
+      undoLatch = new CountDownLatch(TestOperationFails.TOTAL_NUM_OPS);
+      long txid = fate.startTransaction();
+      assertEquals(NEW, getTxStatus(zk, txid));
+      fate.seedTransaction("TestOperationFails", txid,
+          new TestOperationFails(1, ExceptionLocation.CALL), true, "Test Op Fails");
+      assertEquals(SUBMITTED, getTxStatus(zk, txid));
+      // Wait for all the undo() calls to complete
+      undoLatch.await();
+      assertEquals(expectedUndoOrder, TestOperationFails.undoOrder);
+      assertEquals(FAILED_IN_PROGRESS, getTxStatus(zk, txid));
```
Comment (keith-turner) [inline test/src/main/java/org/apache/accumulo/test/fate/zookeeper/FateIT.java:475]:
```suggestion
      assertEquals(FAILED, fate.waitForCompletion(txid));
      assertTrue(fate.getException(txid).getMessage().contains("call() failed"));
```

**Comment by kevinrr888** [comment]
- Removed `SUBMITTED` check in the new test `testRepoFails` (not needed - 
  tested elsewhere)
- Changed tests `testTransactionStatus` and `testCancelWhileInCall` to start
  the transaction runners after the `SUBMITTED` check (the race condition
existed in these tests too)
- Resolved race condition on checking the `FAILED` state of the
  transaction in `testRepoFails`

**Review by keith-turner — APPROVED** [review]

**Pull request merged**